### PR TITLE
Advanced Building Blocks - Enumerables

### PR DESCRIPTION
**Pull request by Oxford-G**
Building Ruby Enumrables

### CODE REVIEW

File: main.rb [hunk 1]
```diff
@@ -0,0 +1,145 @@
+# rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity, Metrics/ModuleLength, Style/For
+module Enumerable
+  # 1.my_each
+  def my_each
+    for i in self
+      yield(i) if block_given?
+    end
+    self
+  end
```
Comment (javitocor) [inline main.rb]: - #my_each returns an Enumerator if no block is given, kindly fix that

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+  # 2. my_each_with_index
+  def my_each_with_index
+    for i in (0..self).size - 1
+      yield(self[i], i) if block_given?
+    end
+    self
+  end
```
Comment (javitocor) [inline main.rb]: - #my_each_with_index should return an enumerator if no block is given
- #my_each_with_index does not mutate the original array
- #my_each_with_index when a block is given when self is an array returns the array itself after calling the given block once for each element in self
- #my_each_with_index when a block is given when self is a range behaves like #each_with_index calls the given block once for each element in self
- #my_each_with_index when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self
- #my_each_with_index when a block is given when self is an array behaves like #each_with_index calls the given block once for each element in self

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+  # 3. my_select
+  def my_select
+    if block_given?
+      result = []
+      my_each do |value|
+        result << value if yield(value)
+      end
+    end
+    result
+  end
```
Comment (javitocor) [inline main.rb]: - #my_select returns an enumerator if no block is given

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+      when Regexp
+        my_each { |x| return true if x =~ parameter }
+      when Class
+        my_each { |x| return true if x.is_a? parameter }
+      else
+        my_each { |element| result = true unless element.nil? }
+      end
+    end
+    result
+  end
```
Comment (javitocor) [inline main.rb]: - #my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    else
+      case parameter
+      when nil
+        size
+      when Numeric
+        my_each { |x| counter += 1 if parameter == x }
+        counter
+      end
+    end
+  end
```
Comment (javitocor) [inline main.rb]: - #my_count counts the number of elements yielding a true value if a block is given
- #my_count does not mutate the original array

**Review by javitocor — CHANGES_REQUESTED** [review]
# STATUS CHANGES REQUIRED 🛑 ♻️ 

Hello! ✋ 
A great effort in this project! 👍 
It is not easy to start working on Ruby with the enumerables! 🤦 
But you did a great job so far! 💯 
However, we still need to make some changes in order to get this project approved! 💪 
Kindly check my comments below 👇  implement the changes and request a new code review ⏩ 
## Mandatory changes
- README file:
    - Project description and built with sections must have the same header size as getting started, author...
    - In the author section, kindly add your LinkedIn profile
    - In the prerequisites section, for this project the user does not need a web browser, you can add instead something like, 'ruby version xxxx'
    - In the getting started section, if you do not want to change that line '$ git clone https://github.com/oxford-g/NYT-webpage-clone.git' make sure the user understands that must be a link similar to this
    - Fill up the about section in the Github page, and do not forget the tags, it will make it easier for the user to know what is the project about

File: main.rb [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+  # 2. my_each_with_index
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    for i in (0..self).size - 1
+      yield(self[i], i)
+    end
+    self
+  end
```
Comment (elmejdki) [inline main.rb]: ## my_each_with_index:

- [ ] You have an error in your my_each_with_index method, please fix it:

image

**Review by elmejdki — CHANGES_REQUESTED** [review]
# status: require changes :recycle: 
## Great job @team :clap: :100: :+1:, Impresive work so far :100:, but we still have some changes :hammer: :wrench: :toolbox: to take care of.


Please check the comments below :point_down: :point_down: :point_down: :point_down: :point_down::

also please run your code and test it before you submit another code reivew.

### Please make all the required changes before you submit another code review.
## Happy Coding :tophat: 
_If you have a question or need help don't hesitate to contact me on slack._

**Review by ElijahTobs — APPROVED** [review]
# Status: Approved! :tada: 

Congratulations! @Oxford-G 

Good job implementing the change as requested by the previous TSE.

## You can now merge and submit.

## Happy Coding :computer: 

_Feel free to reach out to me on slack_ **`@Elijah Ayandokun`**
  **Elijah O. Ayandokun**
[Github](https://github.com/ElijahTobs) | [LinkedIn](https://www.linkedin.com/in/ayandokunelijah/) | [Twitter](https://twitter.com/ElijahTobs)

**Pull request merged**